backends/utils/history_server.go

```
@@ -0,0 +1,363 @@
+)
+
+const (
+	writeMonitoringLogsTimeout = 10 * time.Second
+	pendingLogsBatchSize       = 100
+	logsPathPrefix             = "/monitoring/frames"
+	logsContainer              = "users"
+	maxBytesInNginxRequest     = 5 * 1024 * 1024 // 5MB
+
+	queryType = "query"
+)
+
+type HistoryEntry struct {
+	ActionType     string
+	UserName       string
+	BackendName    string
+	TableName      string
+	Container      string
+	ActionDuration time.Duration
+	StartTime      time.Time
+
+	AdditionalData map[string]string
+}
+
+type HistoryServer struct {
+	logFilePath string
+	logger      logger.Logger
+	logs        []HistoryEntry
+	requests    chan HistoryEntry
+	container   v3io.Container
+	config      *frames.Config
+	isActive    bool
+}
+
+func NewHistoryServer(logger logger.Logger, cfg *frames.Config) *HistoryServer {
+	logPath := fmt.Sprintf("%v/%v_%v.json", logsPathPrefix, uuid.New().String(), time.Now().Unix())
+
+	mon := HistoryServer{logFilePath: logPath,
+		logger:   logger,
+		requests: make(chan HistoryEntry, 100),
+		isActive: false,
+		config:   cfg}
+
+	mon.createDefaultV3ioClient()
+	if mon.isActive {
+		mon.Start()
+	}
+	return &mon
+}
+
+func (m *HistoryServer) Start() {
+	go func() {
+		var pendingLogs []HistoryEntry
+
+		for {
+			select {
+			case entry := <-m.requests:
+				pendingLogs = append(pendingLogs, entry)
+				m.logs = append(m.logs, entry)
+
+				if len(pendingLogs) == pendingLogsBatchSize {
+					m.writeMonitoringBatch(pendingLogs)
+					pendingLogs = pendingLogs[:0]
+				}
+			case <-time.After(writeMonitoringLogsTimeout):
+				if len(pendingLogs) > 0 {
+					m.writeMonitoringBatch(pendingLogs)
+					pendingLogs = pendingLogs[:0]
+				}
+			}
+		}
+	}()
+}
+
+func (m *HistoryServer) createDefaultV3ioClient() {
+
+	session := &frames.Session{}
+
+	token := os.Getenv("V3IO_ACCESS_KEY") // "0938b76f-79ff-41f6-981d-864ea3291cca"
+	if token == "" {
+		m.logger.Warn("can not create v3io.client. could not find `V3IO_ACCESS_KEY` environment variable")
```

So if there's no token we'll just warn about it and keep going?

---

Ok, I see the isActive thing now... Why not report the error though?